### PR TITLE
crc32c: optimize aarch64 crc32c implementation

### PR DESCRIPTION
**Pull request by weixiaowilliam**
ARMv8 defines PMULL crypto instruction.
This patch optimizes crc32c calculate with the instruction when
available rather than original linear crc instructions.

ceph crc32c performance unit test shows that the optimization get
~ x3.90 speedups on ThunderX ARM Core@2.0GHz (Cavium)
~ x1.45 speedups on ARM Cortex-A57@2.1GHz (Huaiwei)
~ x1.16 speedups on ARM Cortex-A57@2.0GHz (Softiron)

Jira: ENTLLT-358
Change-Id: I657422cd20c9ca78237cd060210a5383f4122575

### CODE REVIEW

File: cmake/modules/SIMDExt.cmake [hunk 1]
```diff
@@ -15,34 +15,30 @@
...
-    #define CRC32CX(crc, value) __asm__(\"crc32cx %w[c], %w[c], %x[v]\":[c]\"+r\"(crc):[v]\"r\"(value))
-    asm(\".arch_extension crc\");
-    unsigned int foo(unsigned int ret) {
-      CRC32CX(ret, 0);
-      return ret;
-    }
-    int main() { foo(0); }" HAVE_ARMV8_CRC)
-  set(CMAKE_REQUIRED_QUIET ${save_quiet})
+  set(HAVE_ARM 1)
+  CHECK_C_COMPILER_FLAG(-march=armv8-a+crc HAVE_ARMV8_CRC)
```
Comment (tchaikov) [inline cmake/modules/SIMDExt.cmake]: this basically revert the change of e70ab48b7f6d39a281b3ec65098535a55018b681 by @agraf.

Comment (agraf) [inline cmake/modules/SIMDExt.cmake]: Yes, it does indeed. The rationale for my change was that we're compiling with gcc 4.8, but recent binutils. So -mcpu on the gcc command line doesn't know about +crc.

Comment (agraf) [inline cmake/modules/SIMDExt.cmake]: So a clear NAK from my side :)

Comment (weixiaowilliam) [inline cmake/modules/SIMDExt.cmake]: You are right! we need to consider the compatibility with old gcc. I have refined the patch so that it can switch to original implementation if old gcc doesn't recognize compile flags of "+crc+crypto".

Comment (weixiaowilliam) [inline cmake/modules/SIMDExt.cmake]: You are right! we need to consider the compatibility with old gcc. I have refined the patch so that it can switch to original implementation if old gcc doesn't recognize compile flags of "+crc+crypto".

**Comment by tchaikov** [comment]
@yghannam and @agraf mind taking a look?

File: cmake/modules/SIMDExt.cmake [hunk 2]
```diff
@@ -31,18 +31,22 @@ if(CMAKE_SYSTEM_PROCESSOR MATCHES "aarch64|AARCH64")
   if(HAVE_ARMV8_CRC)
     message(STATUS " aarch64 crc extensions supported")
   endif()
+  CHECK_C_COMPILER_FLAG(-march=armv8-a+crc+crypto HAVE_ARMV8_CRC_CRYPTO_INTRINSICS)
+  if(HAVE_ARMV8_CRC_CRYPTO_INTRINSICS)
+    set(ARMV8_CRC_COMPILE_FLAGS "${ARMV8_CRC_COMPILE_FLAGS} -march=armv8-a+crc+crypto")
```
Comment (agraf) [inline cmake/modules/SIMDExt.cmake]: This checks for gcc flags. Don't do that. Instead, copy the code above and ask the assembler whether it supports crypto extensions.

File: src/common/crc32c_aarch64.c [hunk 3]
```diff
@@ -10,25 +11,123 @@ asm(".arch_extension crc");
 #define CRC32CH(crc, value) __asm__("crc32ch %w[c], %w[c], %w[v]":[c]"+r"(crc):[v]"r"(value))
 #define CRC32CB(crc, value) __asm__("crc32cb %w[c], %w[c], %w[v]":[c]"+r"(crc):[v]"r"(value))
 
+#else
+
+#include <arm_acle.h>
+#include <arm_neon.h>
+
```
Comment (agraf) [inline src/common/crc32c_aarch64.c]: After you changed the check to an assembler based check rather than a compiler based check, please add the an asm(".arch_extension crc+crypto") intrinsic here.

File: cmake/modules/SIMDExt.cmake (hunk 2 above)
Comment (tchaikov) [inline cmake/modules/SIMDExt.cmake]: @agraf but @weixiaowilliam is using the GCC intrinsics to implement the CRC digest. i see it's less portable because lower GCC version might not support these intrinsics, but they are easier to write and easier to read.

@weixiaowilliam maybe we can conditionalize this, by defining `CRC32CX` and `CRC32CW` using assembly (i.e. `crc32w` on aach64) if GCC does not support these intrinsics.

Comment (agraf) [inline cmake/modules/SIMDExt.cmake]: I think we're misunderstanding each other. The problem is that our compiler does not support -march= or -mcpu= on anything but armv8-a.

However, our assembler handles those additional instructions just fine. So all we need to change is the check from the one he's implementing to something that resembles the block above.

Comment (weixiaowilliam) [inline cmake/modules/SIMDExt.cmake]: i have refined the patch by adding inline assembly so that the optimization will work even on lower GCC version which doesn't support crc&crypto intrinsics.  but if on higher GCC version supporting to the intrinsics, it will prefer intrinsics so that ARM compiler has a chance to further optimize it (such as better instruction schedule via more accurate pipeline latency calculation or adopting newest NEON instructions for better performance). besides performance, intrinsics will provide better readability. so the whole logic is now as follows:
```
if compiler supporting crc+crypto intrinsics then
    {optimization based on compiler intrinsics}
else if assembler supporting both crc and crypto instructions then
    {optimization based on crc+crypto instructions}
else if assembler only supporting crc instruction then
    {original optimization based linear crc instructions}
else
    {pure software implementation}
```

**Comment by weixiaowilliam** [comment]
@tchaikov @agraf @yghannam mind taking a look at the refined patch?

**Comment by tchaikov** [comment]
@weixiaowilliam needs rebase

@agraf @yghannam this looks sane to me, do you mind taking a look again?

**Comment by weixiaowilliam** [comment]
@tchaikov rebase is done.

**Comment by weixiaowilliam** [comment]
@agraf @yghannam do you mind taking a look again?
i have refined the patch by adding inline assembly so that the optimization will work even on lower GCC version which doesn't support crc&crypto intrinsics.  but if on higher GCC version supporting to the intrinsics, it will prefer intrinsics so that ARM compiler has a chance to further optimize it (such as better instruction schedule via more accurate pipeline latency calculation or adopting newest NEON instructions for better performance). besides performance, intrinsics will provide better readability. so the whole logic is now as follows:
```
if compiler supporting crc+crypto intrinsics then
    {optimization based on compiler intrinsics}
else if assembler supporting both crc and crypto instructions then
    {optimization based on crc+crypto instructions}
else if assembler only supporting crc instruction then
    {original optimization based linear crc instructions}
else
    {pure software implementation}
```

**Comment by weixiaowilliam** [comment]
@tchaikov @agraf @ @yghannam hi guys, any comments for the aarch64 crc32c optimization?

**Comment by tchaikov** [comment]
@weixiaowilliam adding needs-qa so we can included in next qa batch.

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2017-02-17_10:27:51-rados-wip-kefu-testing---basic-mira/
- http://pulpito.ceph.com/kchai-2017-02-18_07:46:40-rados-wip-kefu-testing---basic-mira/

**Comment by dmick** [comment]
It looks like this requires gcc 4.9.0; CentOS 7.3 appears to include 4.8.5 (and there is no arm_acle.h there).  Researching the best solution.  How much of a difference do the intrinsics make?  Should we just disable them on CentOS/RHEL?

**Comment by dmick** [comment]
it appears as though gcc accepts -march=armv8-a+crc+crypto and yet supplies no arm_acle.h.  I'm not sure what this means.

**Comment by dmick** [comment]
Suggesting a fix like so:  

https://github.com/ceph/ceph/pull/14132
